tesseract_environment/src/environment.cpp

```diff
@@ -1382,37 +1385,156 @@ bool Environment::applyCommandsHelper(const Commands& commands)
+bool Environment::applyAddCommand(const AddLinkCommand::ConstPtr& cmd)
 {
   // The command should not allow this to occur but adding an assert to catch if something changes
   assert(!(!cmd->getLink() && !cmd->getJoint()));
   assert(!((cmd->getLink() != nullptr) && (cmd->getJoint() != nullptr) &&
            (cmd->getJoint()->child_link_name != cmd->getLink()->getName())));
 
+  if (!applyAddLinkCommandHelper(cmd->getLink(), cmd->getJoint(), cmd->replaceAllowed()))
+    return false;
+
+  ++revision_;
+  commands_.push_back(cmd);
+
+  return true;
+}
+
+bool Environment::applyAddTrajectoryLinkCommand(const AddTrajectoryLinkCommand::ConstPtr& cmd)
+{
+  const tesseract_common::JointTrajectory& traj = cmd->getTrajectory();
+  if (traj.empty())
+    return false;
+
+  if (cmd->getLinkName().empty())
+  {
+    CONSOLE_BRIDGE_logWarn("Tried to add trajectory link with empty link name.");
+    return false;
+  }
+
+  if (cmd->getParentLinkName().empty())
+  {
+    CONSOLE_BRIDGE_logWarn("Tried to add trajectory link with empty parent link name.");
+    return false;
+  }
+
+  const bool parent_link_exists = (scene_graph_->getLink(cmd->getParentLinkName()) != nullptr);
+  if (!parent_link_exists)
+  {
+    CONSOLE_BRIDGE_logWarn("Tried to add trajectory link (%s) with parent link (%s) which does not exists.",
+                           cmd->getLinkName().c_str(),
+                           cmd->getParentLinkName().c_str());
+    return false;
+  }
+
+  auto state_solver = state_solver_->clone();
+
+  auto traj_link = std::make_shared<tesseract_scene_graph::Link>(cmd->getLinkName());
+  std::vector<std::string> joint_names;
+  std::vector<std::string> active_link_names;
+  for (const auto& state : traj)
+  {
+    if (state.joint_names.empty())
+    {
+      CONSOLE_BRIDGE_logWarn("Tried to add trajectory link (%s) with empty joint names.", cmd->getLinkName().c_str());
+      return false;
+    }
+
+    if (state.position.rows() == 0)
+    {
+      CONSOLE_BRIDGE_logWarn("Tried to add trajectory link (%s) with empty position.", cmd->getLinkName().c_str());
+      return false;
+    }
+
+    if (state.joint_names.size() != state.position.rows())
+    {
+      CONSOLE_BRIDGE_logWarn("Tried to add trajectory link (%s) where joint names and position are different sizes.",
+                             cmd->getLinkName().c_str());
+      return false;
+    }
+
+    tesseract_scene_graph::SceneState scene_state = state_solver->getState(state.joint_names, state.position);
+    if (joint_names.empty() || !tesseract_common::isIdentical(state.joint_names, joint_names, false))
+    {
+      joint_names = state.joint_names;
+      active_link_names = scene_graph_->getJointChildrenNames(joint_names);
+
+      if (std::find(active_link_names.begin(), active_link_names.end(), cmd->getParentLinkName()) !=
+          active_link_names.end())
+      {
+        CONSOLE_BRIDGE_logWarn("Tried to add trajectory link (%s) where parent link is an active link.",

```

Comment:
Why is that not allowed?

Reply:
Currently when created the joint which attaches the new link it assumes that the parent_link is fixed for all states in the trajectory. Not sure how to make it work if the parent link is allowed to change state to state because it is active.

Reply:
Wait. I just looked at this again. This is only checking if it is an active link of the current trajectory, right? Not global active links of the environment? If so, then that makes sense. I was just thinking if you had 2 arms on a base link that could move  - e.g. a gantry or something